File: clients/swift/src/Services/SpacesService.swift
import Foundation

public final class SpacesService: Sendable {
    private let client: ApiHttpClient

    init(client: ApiHttpClient) { self.client = client }

    public func list() async throws -> [Space] {
        return try await client.send(.init(method: "GET", path: "/spaces"))
    }

    public func create(_ params: CreateSpaceParams) async throws -> Space {
        let body = try client.jsonBody(params)
        return try await client.send(.init(method: "POST", path: "/spaces", body: body, contentType: "application/json"))
    }

    public func get(_ id: Uid) async throws -> Space {
        return try await client.send(.init(method: "GET", path: "/spaces/\(id)"))
    }

    public func update(_ id: Uid, _ params: UpdateSpaceParams) async throws -> Space {
        let body = try client.jsonBody(params)
        return try await client.send(.init(method: "PATCH", path: "/spaces/\(id)", body: body, contentType: "application/json"))
    }

    public func delete(_ id: Uid) async throws {
        try await client.sendVoid(.init(method: "DELETE", path: "/spaces/\(id)"))
    }

    public func addObject(spaceId: Uid, objectId: Uid) async throws {
        try await client.sendVoid(.init(method: "PUT", path: "/spaces/\(spaceId)/objects/\(objectId)"))
    }

    public func removeObject(spaceId: Uid, objectId: Uid) async throws {
        try await client.sendVoid(.init(method: "DELETE", path: "/spaces/\(spaceId)/objects/\(objectId)"))
    }
}

File: clients/typescript/src/services/SpacesService.ts
import type { CreateSpaceRequest, UpdateSpaceRequest } from "../actions/index.js";
import type { ApiClient, RequestOptions } from "../client.js";
import type { Space } from "../models/index.js";

/**
 * Resource methods for `/spaces`. See https://access.mymind.com/api → Spaces.
 */
export class SpacesService {
  constructor(private readonly client: ApiClient) {}

  /** `GET /spaces`. */
  async list(options?: RequestOptions): Promise<Space[]> {
    return this.client.send<Space[]>({
      method: "GET",
      path: "/spaces",
      signal: options?.signal,
    });
  }

  /** `POST /spaces`. */
  async create(
    request: CreateSpaceRequest,
    options?: RequestOptions,
  ): Promise<Space> {
    return this.client.send<Space>({
      method: "POST",
      path: "/spaces",
      body: request,
      signal: options?.signal,
    });
  }

  /** `GET /spaces/:id`. */
  async get(id: string, options?: RequestOptions): Promise<Space> {
    return this.client.send<Space>({
      method: "GET",
      path: `/spaces/${id}`,
      signal: options?.signal,
    });
  }

  /** `PATCH /spaces/:id`. */
  async update(
    id: string,
    request: UpdateSpaceRequest,
    options?: RequestOptions,
  ): Promise<Space> {
    return this.client.send<Space>({
      method: "PATCH",
      path: `/spaces/${id}`,
      body: request,
      signal: options?.signal,
    });
  }

  /** `DELETE /spaces/:id`. Idempotent. */
  async delete(id: string, options?: RequestOptions): Promise<void> {
    await this.client.send<void>({
      method: "DELETE",
      path: `/spaces/${id}`,
      signal: options?.signal,
    });
  }

  /** `PUT /spaces/:spaceId/objects/:objectId` — add object to space. Idempotent. */
  async addObject(
    spaceId: string,
    objectId: string,
    options?: RequestOptions,
  ): Promise<void> {
    await this.client.send<void>({
      method: "PUT",
      path: `/spaces/${spaceId}/objects/${objectId}`,
      signal: options?.signal,
    });
  }

  /** `DELETE /spaces/:spaceId/objects/:objectId`. Idempotent. */
  async removeObject(
    spaceId: string,
    objectId: string,
    options?: RequestOptions,
  ): Promise<void> {
    await this.client.send<void>({
      method: "DELETE",
      path: `/spaces/${spaceId}/objects/${objectId}`,
      signal: options?.signal,
    });
  }
}
